File: 461.汉明距离.java
/*
 * @lc app=leetcode.cn id=461 lang=java
 *
 * [461] 汉明距离
 */

// @lc code=start
class Solution {
    public int hammingDistance(int x, int y) {
        int xor = x ^ y, ans = 0;
        // count the number of bit 1's in XOR(x, y)
        while (xor > 0) {
            // Brian Kernighan Algorithm: erases the rightmost bit 1
            xor &= xor - 1;
            ++ans;
        }
        return ans;
    }
}
// @lc code=end

File: 461.汉明距离.py
#
# @lc app=leetcode.cn id=461 lang=python3
#
# [461] 汉明距离
#

# @lc code=start
class Solution:
    def hammingDistance(self, x: int, y: int) -> int:
        xor = x ^ y
        dist = 0
        while xor:
            # mask out the rest bits
            # if xor & 1:
            dist += (xor & 1)

            # shift bits to the right (divided by 2) until 
            # all non-zero bits are counted
            xor >>= 1
        return dist

# @lc code=end
